renderdoc/driver/d3d12/d3d12_counters.cpp

```diff
@@ -228,106 +241,257 @@ vector<CounterResult> D3D12Replay::FetchCounters(const vector<uint32_t> &counter
   }
 
-  D3D12_QUERY_HEAP_DESC queryDesc;
-  queryDesc.Count = maxEID * 2;
-  queryDesc.NodeMask = 1;
-  queryDesc.Type = D3D12_QUERY_HEAP_TYPE_TIMESTAMP;
-  ID3D12QueryHeap *queryHeap = NULL;
-  hr = m_pDevice->CreateQueryHeap(&queryDesc, __uuidof(queryHeap), (void **)&queryHeap);
+  D3D12_RESOURCE_DESC occlusionBufDesc;
+  occlusionBufDesc.Dimension = D3D12_RESOURCE_DIMENSION_BUFFER;
+  occlusionBufDesc.Alignment = 0;
+  occlusionBufDesc.Width = sizeof(uint64_t) * maxEID;
+  occlusionBufDesc.Height = 1;
+  occlusionBufDesc.DepthOrArraySize = 1;
+  occlusionBufDesc.MipLevels = 1;
+  occlusionBufDesc.Format = DXGI_FORMAT_UNKNOWN;
+  occlusionBufDesc.SampleDesc.Count = 1;
+  occlusionBufDesc.SampleDesc.Quality = 0;
+  occlusionBufDesc.Layout = D3D12_TEXTURE_LAYOUT_ROW_MAJOR;
+  occlusionBufDesc.Flags = D3D12_RESOURCE_FLAG_NONE;
+
+  ID3D12Resource *occlusionReadbackBuf;
+  hr = m_pDevice->CreateCommittedResource(&heapProps, D3D12_HEAP_FLAG_NONE, &occlusionBufDesc,
+                                          D3D12_RESOURCE_STATE_COPY_DEST, NULL,
+                                          __uuidof(ID3D12Resource), (void **)&occlusionReadbackBuf);
   if(FAILED(hr))
   {
-    RDCERR("Failed to create query heap %08x", hr);
+    RDCERR("Failed to create occlusion query readback buffer %08x", hr);
     return ret;
   }
 
-  m_pDevice->SetStablePowerState(TRUE);
+  D3D12_QUERY_HEAP_DESC timerQueryDesc;
+  timerQueryDesc.Count = maxEID * 2;
+  timerQueryDesc.NodeMask = 1;
+  timerQueryDesc.Type = D3D12_QUERY_HEAP_TYPE_TIMESTAMP;
+  ID3D12QueryHeap *timerQueryHeap = NULL;
+  hr = m_pDevice->CreateQueryHeap(&timerQueryDesc, __uuidof(timerQueryHeap),
+                                  (void **)&timerQueryHeap);
+  if(FAILED(hr))
+  {
+    RDCERR("Failed to create timer query heap %08x", hr);
+    return ret;
+  }
 
-  D3D12GPUTimerCallback cb(m_pDevice, this, queryHeap);
+  D3D12_QUERY_HEAP_DESC pipestatsQueryDesc;
+  pipestatsQueryDesc.Count = maxEID;
+  pipestatsQueryDesc.NodeMask = 1;
+  pipestatsQueryDesc.Type = D3D12_QUERY_HEAP_TYPE_PIPELINE_STATISTICS;
+  ID3D12QueryHeap *pipestatsQueryHeap = NULL;
+  hr = m_pDevice->CreateQueryHeap(&pipestatsQueryDesc, __uuidof(pipestatsQueryHeap),
+                                  (void **)&pipestatsQueryHeap);
+  if(FAILED(hr))
+  {
+    RDCERR("Failed to create pipeline statistics query heap %08x", hr);
+    return ret;
+  }
+
+  D3D12_QUERY_HEAP_DESC occlusionQueryDesc;
+  occlusionQueryDesc.Count = maxEID;
+  occlusionQueryDesc.NodeMask = 1;
+  occlusionQueryDesc.Type = D3D12_QUERY_HEAP_TYPE_OCCLUSION;
+  ID3D12QueryHeap *occlusionQueryHeap = NULL;
+  hr = m_pDevice->CreateQueryHeap(&occlusionQueryDesc, __uuidof(occlusionQueryHeap),
+                                  (void **)&occlusionQueryHeap);
+  if(FAILED(hr))
+  {
+    RDCERR("Failed to create occlusion query heap %08x", hr);
+    return ret;
+  }
+
+  // Only supported with developer mode drivers!!!
+  // m_pDevice->SetStablePowerState(TRUE);
```

Comment:
I don't think we can comment this out, it's important to try and get reliable results on some implementations. As I mentioned in #457, the fact that this kills the device when some switch isn't flipped is a flaw in D3D12, but we can't ignore it.

What I'd suggest is check the HRESULT returned, and if it failed with DEVICE_REMOVED then use `MessageBoxA` to print a message to the user, something like "Fatal error: SetStablePowerState failed, enable [[whatver magic doohicky]] to use this feature.". Obviously the replay program might crash after that, but it's not worth trying to add graceful shutdown code to workaround an OS bug.